src/cdk-experimental/menu/background-click-service.ts

```diff
@@ -0,0 +1,129 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Injectable, Inject, InjectionToken, OnDestroy} from '@angular/core';
+import {DOCUMENT} from '@angular/common';
+import {fromEvent, Subject} from 'rxjs';
+import {takeUntil} from 'rxjs/operators';
+import {MenuStack} from './menu-stack';
+
+/**
+ * A function which decides whether the background click event should cause the menu stack to close
+ * out. It is provided alongside the MenuStack to the BackgroundClickService and is called on each
+ * background click event.
+ */
+export type CloseDecider = (target: Element | null) => boolean;
+
+/** Injection token for CloseDecider function. */
+export const CLOSE_DECIDER = new InjectionToken<CloseDecider>('cdk-menu-close-decider');
+
+/**
+ * BackgroundClickService listens for background click events when started and performs menu
+ * closing actions. When the service detects a background click, it calls upon the CloseDecider
+ * function in order to determine if the current MenuStack should be closed out.
+ *
+ * Note that it is acceptable to start the listener from nested triggers as new listeners will not
+ * be added for the same MenuStack. If the service is started with a new MenuStack, the previous
+ * set of menus is closed out resulting in only having a single set of open menus at any given time.
+ */
+@Injectable({providedIn: 'root'})
+export class BackgroundClickService implements OnDestroy {
```

Comment:
I don't totally follow why this service is necessary. Why can't we use the `outsidePointerEvents` stream on `OverlayRef` for most of this, filtering the stream to the events that we care about for each type of overlay? I tried doing roughly what I had in mind here: https://github.com/jelbourn/components/commit/584265c7840959dff7f1c1b2ca0dce78b30d12eb

What am I missing with this approach?

Reply:
Agreed to update the Overlay logic to listen to `contextmenu` events (#20373) and to exclude buttons with a `cdk-menu-trigger` class from the predicates.